nginx/modsecurity/ngx_http_modsecurity.c

```diff
@@ -848,19 +898,81 @@ ngx_http_modsecurity_body_handler(ngx_http_request_t *r)
     ngx_http_core_run_phases(r);
 }
 
+static ngx_int_t
+ngx_http_calculate_modsec_latency(ngx_http_request_t *r, ngx_http_modsecurity_ctx_t* ctx,
+                                  ngx_http_modsecurity_loc_conf_t * cf)
+{
+    ngx_time_t *end_time;
+    ngx_msec_int_t   ms;
+    ngx_int_t ret;
+    ngx_time_update();
+    end_time = ngx_timeofday();
+
+    ms = (ngx_msec_int_t)
+        ((end_time->sec - ctx->start_time->sec) * 1000 + (end_time->msec - ctx->start_time->msec));
+    ms = ngx_max(ms, 0);
+
+    ngx_http_variable_value_t* modsec_latency_detect_var = ngx_http_get_indexed_variable(r, modsec_latency_detect_index);
+    ngx_http_variable_value_t* modsec_latency_prevent_var = ngx_http_get_indexed_variable(r, modsec_latency_prevent_index);
+    if (cf->config->is_enabled == MODSEC_DETECTION_ONLY) {
+        ret = ngx_http_modsecurity_set_modsec_latency(r, modsec_latency_detect_var, ms);
+        if (ret != NGX_OK) {
+            return ret;
+        }
+        return ngx_http_modsecurity_set_modsec_latency(r, modsec_latency_prevent_var, 0);
```

Comment:
Is it a valid scenario that the  ngx_pnalloc inside the set_modsec_latency for the first call succeeds and the second call fails? Are we considering this scenario?

Reply:
yes we have considered this scenario. Only in the first call the memory allocation happens and in future calls we check if v->data == NULL which would be false and hence future allocations wont happen.